crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,966 @@
+    const fn new() -> Self {
+        Self([const { Chunk::new() }; Self::NUM_CHUNKS as usize])
+    }
+
+    /// Computes the capacity of the chunk at this index within [`Self::NUM_CHUNKS`].
+    /// The first 2 have length 512 (2^9) and the last has length (2^31)
+    #[inline]
+    fn capacity_of_chunk(chunk_index: u32) -> u32 {
+        // We do this because we're skipping the first `NUM_SKIPPED` powers, so we need to make up for them by doubling the first index.
+        // This is why the first 2 indices both have a capacity of 512.
+        let corrected = chunk_index.max(1);
+        // We add NUM_SKIPPED because the total capacity should be as if [`Self::NUM_CHUNKS`] were 32.
+        // This skips the first NUM_SKIPPED powers.
+        let corrected = corrected + Self::NUM_SKIPPED;
+        // This bit shift is just 2^corrected.
+        1 << corrected
+    }
+
+    /// For this index in the whole buffer, returns the index of the [`Chunk`], the index within that chunk, and the capacity of that chunk.
+    #[inline]
+    fn index_info(full_index: u32) -> (u32, u32, u32) {
+        // We do a `saturating_sub` because we skip the first `NUM_SKIPPED` powers to make space for the first chunk's entity count.
+        // The -1 is because this is the number of chunks, but we want the index in the end.
+        // We store chunks in smallest to biggest order, so we need to reverse it.
+        let chunk_index = (Self::NUM_CHUNKS - 1).saturating_sub(full_index.leading_zeros());
+        let chunk_capacity = Self::capacity_of_chunk(chunk_index);
+        // We only need to cut off this particular bit.
+        // The capacity is only one bit, and if other bits needed to be dropped, `leading` would have been greater
+        let index_in_chunk = full_index & !chunk_capacity;
+
+        (chunk_index, index_in_chunk, chunk_capacity)
+    }
+
+    /// For this index in the whole buffer, returns the [`Chunk`], the index within that chunk, and the capacity of that chunk.
+    #[inline]
+    fn index_in_chunk(&self, full_index: u32) -> (&Chunk, u32, u32) {
+        let (chunk_index, index_in_chunk, chunk_capacity) = Self::index_info(full_index);
+        // SAFETY: Caller ensures the chunk index is correct
+        let chunk = unsafe { self.0.get_unchecked(chunk_index as usize) };
+        (chunk, index_in_chunk, chunk_capacity)
+    }
+
+    /// Gets the entity at an index.
+    ///
+    /// # Safety
+    ///
+    /// [`set`](Self::set) must have been called on this index to initialize the its memory.
+    unsafe fn get(&self, full_index: u32) -> Entity {
+        let (chunk, index, _) = self.index_in_chunk(full_index);
+        // SAFETY: Caller ensures this index was set
+        unsafe { chunk.get(index) }
+    }
+
+    /// Sets an entity at an index.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently with itself.
+    /// Access must not conflict with another [`Self::get`].
+    #[inline]
```

Comment:
It can't be called concurrently with itself *at all*? It looks like some level of concurrency is fine, but that sets and gets must be ordered *for a given index*.